src/dataflow/src/render/reduce.rs

```diff
@@ -1212,66 +1353,109 @@ where
+                                with non-zero accumulation: {:?}: {:?}",
+                                aggr,
+                                sum
+                            );
+                        }
+                        _ => (),
                     }
 
                     // The finished value depends on the aggregation function in a variety of ways.
-                    let value = match (&aggr.func, agg2) {
-                        (AggregateFunc::Count, _) => Datum::Int64(agg2 as i64),
-                        (AggregateFunc::All, _) => {
+                    let value = match (&aggr.func, accum) {
+                        (
+                            AggregateFunc::Count,
+                            Accumulator::Count {
+                                non_nulls,
+                                total: _,
+                            },
+                        ) => Datum::Int64(i64::cast_from(*non_nulls)),
+                        (
+                            AggregateFunc::All,
+                            Accumulator::Bool {
+                                falses,
+                                trues,
+                                total,
+                            },
+                        ) => {
                             // If any false, else if all true, else must be no false and some nulls.
-                            if agg2 > 0 {
+                            if *falses > 0 {
                                 Datum::False
-                            } else if tot == agg1 {
+                            } else if trues == total {
                                 Datum::True
                             } else {
                                 Datum::Null
                             }
                         }
-                        (AggregateFunc::Any, _) => {
+                        (
+                            AggregateFunc::Any,
+                            Accumulator::Bool {
+                                falses,
+                                trues,
+                                total,
+                            },
+                        ) => {
                             // If any true, else if all false, else must be no true and some nulls.
-                            if agg1 > 0 {
+                            if *trues > 0 {
                                 Datum::True
-                            } else if tot == agg2 {
+                            } else if falses == total {
                                 Datum::False
                             } else {
                                 Datum::Null
                             }
                         }
                         (AggregateFunc::Dummy, _) => Datum::Dummy,
                         // Below this point, anything with only nulls should be null.
-                        (_, 0) => Datum::Null,
+                        (_, Accumulator::SimpleNumber { sum: _, count: 0 }) => Datum::Null,
```

Comment:
I think this would be very surprising to see, as I think it means that we have a net-zero number of records, and a non-zero `sum` would be an error (and a zero sum should be a zero accumulation and not presented to us).

Reply:
Good point. This is refactored now per discussion above.